docs/patterns.md

```diff
@@ -142,3 +142,132 @@ Example access policy:
 * The front desk cannot cut releases nor deploy services
 
 Complex policies like this are currently best implemented in code directly, though there is [ongoing work](https://github.com/michaelansel/hubot-rbac) to build a generalized framework for access management.
+
+## Middleware Examples
+
+Middleware is often used for authentication and authorization.  There are three kinds of middleware:
```

Comment:
Okay, once we get the interfaces figured out I'll reorganize to put some of the explanation in scripting.md and then use cases in patterns.md.